kernel/src/process.rs

```diff
@@ -1212,6 +1255,125 @@ impl<C: Chip> ProcessType for Process<'_, C> {
+
+        // A callback is identified by a tuple of the driver number
+        // and the subscribe number
+        let callback_id = CallbackId {
+            driver_num,
+            subscribe_num,
+        };
+
+        // Only one callback should exist per tuple, to ensure that
+        // there are no pending callbacks with the same identifier but
+        // an old function pointer, clear them now.
+        //
+        // This operates only on the current process, so the passed
+        // pointer does not need to be trusted.
+        self.remove_pending_callbacks(callback_id);
+
+        // Construct the callback struct
+        let callback = Callback::new(self.appid(), callback_id, appdata, fn_ptr);
+
+        // Invoke the capsule
+        let driver_res = driver_invoc_closure(callback);
+
+        match driver_res {
+            Err((returned_callback, err)) => {
+                // The capsule has refused the subscribe operation,
+                // verify that it passed back the new callback
+                if let Some(process_callback) = returned_callback.into_inner() {
+                    if process_callback.app_id != self.appid()
+                        || process_callback.callback_id != callback_id
+                        || process_callback.appdata != appdata
+                        || process_callback.fn_ptr != fn_ptr
+                    {
+                        // The capsule did not return the Callback passed in
+                        // TODO: How to handle this?
+                        panic!(
+                            "Driver {}, subscribe num {}: Callback swapped in error case",
+                            driver_num, subscribe_num
+                        );
+                    } else {
+                        // Capsule returned the correct Callback
+                        GenericSyscallReturnValue::SubscribeFailure(err, callback_ptr, appdata)
+                    }
+                } else {
+                    // The capsule returned a default Callback
+                    // instance in the error-case return value, where
+                    // it should have passed the argument back
+                    panic!(
+                        "Driver {}, subcribe_num {}: Default callback returned in error case",
+                        driver_num, subscribe_num
+                    );
+                }
+            }
+            Ok(returned_callback) => {
+                // The capsule indicated that the subscribe operation succeeded
+                match returned_callback.into_inner() {
+                    None => {
+                        // The capsule returned a default callback, we
+                        // have to believe this to be correct
+                        //
+                        // TODO: Limit the default constructor on
+                        // Callback to require a capability for Grant
+                        // initializations
```

Comment:
I don't see how using capabilities to limit calls to the default constructor can prevent capsules from storing callbacks across multiple `subscribe` calls.

Consider a capsule that contains a `[Callback; 2]`. The first time `subscribe` is called, it swaps the app's callback with the first entry in the array. The second time, it swaps the app's callback with the second entry in the array. Now the capsule has both the old and new callback, and the app is unaware the callback is stored.

Reply:
I suppose you're right, there's always going to be a way for a capsule to hold onto `Callback`s for capsules, if they are allowed to return a default Callback. I was also concerned on how to create default instances requiring a capability for `Callback`s not stored in Grants, that would've required passing them in from the `main.rs`.